File: packages/js/src/sign.ts
/**
 * HAP claim signing functions (for Verification Authorities)
 */

import * as jose from "jose";
import type { KeyLike } from "jose";
import { VERSION, Claim, ClaimCost, SignOptions } from "./types";

/** Characters used for ID generation */
const ID_CHARS =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

/**
 * Generates a cryptographically secure random HAP ID
 * @returns A HAP ID in the format hap_[a-zA-Z0-9]{12}
 */
export function generateId(): string {
  const randomBytes = crypto.getRandomValues(new Uint8Array(12));
  let suffix = "";

  for (let i = 0; i < 12; i++) {
    suffix += ID_CHARS[randomBytes[i] % ID_CHARS.length];
  }

  return `hap_${suffix}`;
}

/**
 * Generates a test HAP ID (for previews and development)
 * @returns A test HAP ID in the format hap_test_[a-zA-Z0-9]{8}
 */
export function generateTestId(): string {
  const randomBytes = crypto.getRandomValues(new Uint8Array(8));
  let suffix = "";

  for (let i = 0; i < 8; i++) {
    suffix += ID_CHARS[randomBytes[i] % ID_CHARS.length];
  }

  return `hap_test_${suffix}`;
}

/**
 * Checks if a HAP ID is a test ID
 * @param id - The HAP ID to check
 * @returns true if the ID is a test ID
 */
export function isTestId(id: string): boolean {
  return /^hap_test_[a-zA-Z0-9]{8}$/.test(id);
}

/**
 * Generates a new Ed25519 key pair for signing HAP claims
 * @returns Object containing the public and private keys
 */
export async function generateKeyPair(): Promise<{
  publicKey: KeyLike;
  privateKey: KeyLike;
}> {
  const { publicKey, privateKey } = await jose.generateKeyPair("EdDSA", {
    crv: "Ed25519",
  });

  return { publicKey, privateKey };
}

/**
 * Exports a public key to JWK format suitable for /.well-known/hap.json
 * @param publicKey - The public key to export
 * @param kid - The key ID to assign
 * @returns JWK object
 */
export async function exportPublicKeyJwk(
  publicKey: KeyLike,
  kid: string,
): Promise<jose.JWK> {
  const jwk = await jose.exportJWK(publicKey);
  return {
    ...jwk,
    kid,
    kty: "OKP",
    crv: "Ed25519",
  };
}

/**
 * Signs a HAP claim with an Ed25519 private key
 * @param claim - The claim to sign
 * @param privateKey - The Ed25519 private key
 * @param options - Signing options (must include kid)
 * @returns JWS compact serialization string
 */
export async function signClaim(
  claim: Claim,
  privateKey: KeyLike,
  options: SignOptions,
): Promise<string> {
  // Ensure version is set
  const claimWithVersion = {
    ...claim,
    v: claim.v || VERSION,
  };

  const jws = await new jose.CompactSign(
    new TextEncoder().encode(JSON.stringify(claimWithVersion)),
  )
    .setProtectedHeader({
      alg: "EdDSA",
      kid: options.kid,
    })
    .sign(privateKey);

  return jws;
}

/**
 * Parameters for creating a HAP claim
 */
export interface CreateClaimParams {
  /** VA-specific verification method identifier */
  method: string;
  /** Human-readable description of the effort */
  description: string;
  /** Recipient name */
  recipientName: string;
  /** Recipient domain (optional) */
  domain?: string;
  /** Issuer domain (VA's domain) */
  issuer: string;
  /** Service tier (optional) */
  tier?: string;
  /** Days until expiration (optional) */
  expiresInDays?: number;

  // Effort dimensions (all optional)

  /** Monetary cost incurred */
  cost?: ClaimCost;
  /** Time in seconds of exclusive dedication */
  time?: number;
  /** Whether physical-world atoms were involved */
  physical?: boolean;
  /** Human energy expenditure in kilocalories */
  energy?: number;
}

/**
 * Creates a complete HAP claim with all required fields
 * @param params - Claim parameters
 * @returns A complete Claim object
 */
export function createClaim(params: CreateClaimParams): Claim {
  const now = new Date();
  const claim: Claim = {
    v: VERSION,
    id: generateId(),
    to: {
      name: params.recipientName,
      ...(params.domain && { domain: params.domain }),
    },
    at: now.toISOString(),
    iss: params.issuer,
    method: params.method,
    description: params.description,
  };

  if (params.tier) {
    claim.tier = params.tier;
  }

  if (params.expiresInDays) {
    const exp = new Date(now);
    exp.setDate(exp.getDate() + params.expiresInDays);
    claim.exp = exp.toISOString();
  }

  // Add effort dimensions if provided
  if (params.cost) {
    claim.cost = params.cost;
  }

  if (params.time !== undefined) {
    claim.time = params.time;
  }

  if (params.physical !== undefined) {
    claim.physical = params.physical;
  }

  if (params.energy !== undefined) {
    claim.energy = params.energy;
  }

  return claim;
}

/**
 * Computes SHA-256 hash of content with prefix
 * @param content - The content to hash
 * @returns Hash string in format "sha256:xxxxx"
 */
export async function hashContent(content: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(content);
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
  return `sha256:${hashHex}`;
}

File: packages/go/sign.go
package humanattestation

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"

	"github.com/go-jose/go-jose/v4"
)

// IDChars contains characters used for HAP ID generation
const IDChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

// GenerateID generates a cryptographically secure random HAP ID
func GenerateID() (string, error) {
	bytes := make([]byte, 12)
	if _, err := rand.Read(bytes); err != nil {
		return "", fmt.Errorf("failed to generate random bytes: %w", err)
	}

	suffix := make([]byte, 12)
	for i := 0; i < 12; i++ {
		suffix[i] = IDChars[int(bytes[i])%len(IDChars)]
	}

	return "hap_" + string(suffix), nil
}

// GenerateTestID generates a test HAP ID (for previews and development)
func GenerateTestID() (string, error) {
	bytes := make([]byte, 8)
	if _, err := rand.Read(bytes); err != nil {
		return "", fmt.Errorf("failed to generate random bytes: %w", err)
	}

	suffix := make([]byte, 8)
	for i := 0; i < 8; i++ {
		suffix[i] = IDChars[int(bytes[i])%len(IDChars)]
	}

	return "hap_test_" + string(suffix), nil
}

// IsTestID checks if a HAP ID is a test ID
func IsTestID(id string) bool {
	return TestIDRegex.MatchString(id)
}

// HashContent computes SHA-256 hash of content with prefix
func HashContent(content string) string {
	hash := sha256.Sum256([]byte(content))
	return "sha256:" + hex.EncodeToString(hash[:])
}

// GenerateKeyPair generates a new Ed25519 key pair for signing HAP claims
func GenerateKeyPair() (ed25519.PrivateKey, ed25519.PublicKey, error) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to generate key pair: %w", err)
	}
	return privateKey, publicKey, nil
}

// ExportPublicKeyJWK exports a public key to JWK format suitable for /.well-known/hap.json
func ExportPublicKeyJWK(publicKey ed25519.PublicKey, kid string) JWK {
	x := base64.RawURLEncoding.EncodeToString(publicKey)
	return JWK{
		Kid: kid,
		Kty: "OKP",
		Crv: "Ed25519",
		X:   x,
	}
}

// SignClaim signs a HAP claim with an Ed25519 private key
func SignClaim(claim *Claim, privateKey ed25519.PrivateKey, kid string) (string, error) {
	// Serialize the claim
	payload, err := json.Marshal(claim)
	if err != nil {
		return "", fmt.Errorf("failed to serialize claim: %w", err)
	}

	// Create the signer
	signer, err := jose.NewSigner(
		jose.SigningKey{Algorithm: jose.EdDSA, Key: privateKey},
		(&jose.SignerOptions{}).WithHeader("kid", kid),
	)
	if err != nil {
		return "", fmt.Errorf("failed to create signer: %w", err)
	}

	// Sign the payload
	jws, err := signer.Sign(payload)
	if err != nil {
		return "", fmt.Errorf("failed to sign claim: %w", err)
	}

	// Serialize to compact format
	compact, err := jws.CompactSerialize()
	if err != nil {
		return "", fmt.Errorf("failed to serialize JWS: %w", err)
	}

	return compact, nil
}

// CreateClaimParams contains parameters for creating a HAP claim
type CreateClaimParams struct {
	Method        string
	Description   string
	RecipientName string
	Domain        string
	Tier          string
	Issuer        string
	ExpiresInDays int
	Cost          *ClaimCost
	Time          *int
	Physical      *bool
	Energy        *int
}

// CreateClaim creates a complete HAP claim with all required fields
func CreateClaim(params CreateClaimParams) (*Claim, error) {
	id, err := GenerateID()
	if err != nil {
		return nil, err
	}

	now := time.Now().UTC()
	claim := &Claim{
		V:           Version,
		ID:          id,
		Method:      params.Method,
		Description: params.Description,
		To: ClaimTarget{
			Name:   params.RecipientName,
			Domain: params.Domain,
		},
		At:  now.Format(time.RFC3339),
		Iss: params.Issuer,
	}

	if params.Tier != "" {
		claim.Tier = params.Tier
	}

	if params.ExpiresInDays > 0 {
		exp := now.AddDate(0, 0, params.ExpiresInDays)
		claim.Exp = exp.Format(time.RFC3339)
	}

	// Add effort dimensions if provided
	if params.Cost != nil {
		claim.Cost = params.Cost
	}

	if params.Time != nil {
		claim.Time = params.Time
	}

	if params.Physical != nil {
		claim.Physical = params.Physical
	}

	if params.Energy != nil {
		claim.Energy = params.Energy
	}

	return claim, nil
}
